### DISCUSSION TITLE
System.DllNotFoundException: 'Unable to load DLL 'mmdeploy' or one of its dependencies: The specified module could not be found. (0x8007007E)'

### DISCUSSION THREAD

**Discussion by pd2871** [open-mmlab/mmdeploy#1730] [Q&A]
@irexyc , thanks for your help in my last question.

After building and installing the csharp `demo.sln`, when I tried to run `image_segmentation`, it throwed the error `System.DllNotFoundException: 'Unable to load DLL 'mmdeploy' or one of its dependencies: The specified module could not be found. (0x8007007E)'`, I build everything and all went fine without any error, What can be possible error for this? Thanks
 
![Screenshot (2802)](https://user-images.githubusercontent.com/59336638/217445511-26f60aed-c85b-4ee3-8513-919ed108f15f.png)

**Comment by irexyc** [comment]
HI @pd2871 , which mmdeploy version are you using? We change `MMDEPLOY_SHARED_LIBS` default to `OFF` recently. Therefore, after build mmdeploy, there will only one dll called `mmdeploy.dll` (except xxx_ops.dll).

There are some dlls that `mmdeploy.dll` depends on. Like dlls for backends lib (cudnn, tensorrt, onnxruntime etc.) and opencv (if you didn't use opencv static lib) which need added to the` PATH` You can use this [tools](https://github.com/lucasg/Dependencies) to check if the dependencies of `mmdeploy.dll` satisfied on you environment.

What needs to be specified is that if you use onnxruntime, there exists two onnxruntime.dll in the system dir that may be loaded first . I have no clue what these dlls are used for. To make mmdeploy run with onnxruntime, we need to hide these two dlls. One way is to change names of there two dlls (better than delete for it can be restored). As the executable will first try to find dependencies from its own path,  another way is to copy the right onnxruntime related dlls to the executable dir.

```
C:\Windows\SysWOW64\onnxruntime.dll
C:\Windows\System32\onnxruntime.dll
```

**Comment by pd2871** [comment]
@irexyc  Thanks for your suggestions, I will apply it and get back to you

**Reply by pd2871** [reply]
@irexyc  I can see 2 folders inside `mmdeploy\apis\csharp\MMDeploy\bin\Release`, those 2 folders both have `mmdeploy.dll`, can it be the reason?


![Screenshot (2803)](https://user-images.githubusercontent.com/59336638/217461939-3741644f-0d50-4981-b0d7-fb3071a72a47.png)
![Screenshot (2804)](https://user-images.githubusercontent.com/59336638/217462602-edf49a9d-8a35-4849-8698-7231424935a8.png)

**Comment by irexyc** [comment]
@pd2871 
When you build nuget package, it will generate two different version as net48 and netcore3.1. And when you use it in a project, it will install one of them according to you project setting.

From the picture you paste, the `mmdeploy.dll` is 1801 KB. According to the small size, I think you build mmdeploy with `MMDEPLOY_SHARED_LIBS=ON`.

You could have a look as mmdeploy\build\bin\Release, if there are multi dlls like `mmdeploy_core.dll`, `mmdeploy_cpu_device` ...
These dlls are dll dependencies for `mmdeploy.dll`. You should paste these dlls to the demo project path where mmdeploy.dll located. While this way is not recommanded, you'd better build mmdeploy with `MMDEPLOY_SHARED_LIBS=OFF`. (static opencv lib is prefered to reduce the dependencies of mmdeploy.dll)

**Reply by pd2871** [reply]
Hi @irexyc , I checked the directory and it contained only 2 `dlls`, also I checked the tool that you have shared to check the dependencies, You can see the images below. Also, should I need to build the `demo.sln` in `build` mode or `release` mode?

![Screenshot (2805)](https://user-images.githubusercontent.com/59336638/217478327-fa05ad1d-5192-4365-a0ab-7ff8808cd061.png)
![Screenshot (2806)](https://user-images.githubusercontent.com/59336638/217478530-25d73938-9a0b-46bc-a919-c031a1c12ffa.png)

**Reply by pd2871** [reply]
Also, this the command that I used to build:
```
cd $env:MMDEPLOY_DIR
mkdir build -ErrorAction SilentlyContinue
cd build
cmake .. -G "Visual Studio 16 2019" -A x64 -T v142 `
    -DMMDEPLOY_BUILD_SDK=ON `
    -DOpenCV_DIR=C:\onnx_test\opencv\build\x64\vc16\lib `
    -DMMDEPLOY_BUILD_EXAMPLES=ON `
    -DMMDEPLOY_BUILD_SDK_CSHARP_API=ON `
    -DMMDEPLOY_TARGET_DEVICES="cpu" `
    -DMMDEPLOY_TARGET_BACKENDS="ort" `
    -DONNXRUNTIME_DIR="$env:ONNXRUNTIME_DIR"

cmake --build . --config Release -- /m
cmake --install . --config Release

```

**Comment by pd2871** [comment]
Hi @irexyc , this [tool](https://github.com/lucasg/Dependencies) that you have shared was very helpful, I figured out ONNX and OpenCV build failed as in the image above, I replaced the old `onnxruntime.dll` of `C:\\Windows\System32` with the new one from `onnxruntime-win-x64-1.8.1\lib\onnxruntime.dll` and for OpenCV, I changed to the actual folder of OpenCV build location and after rebuilding the problem solved.
![Screenshot (2808)](https://user-images.githubusercontent.com/59336638/217495124-b5a22307-b44b-4ffd-aa92-3a3226faf64c.png)
